### PR TITLE
Add `out` argument to `ndarray.get` for asynchronous device-to-host copy

### PR DESCRIPTION
**Pull request by okuta**
This PR is takeover version of #1955.
- Reduce function call
- Fix flake8 error
- Add tests

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1621,30 +1621,63 @@ cdef class ndarray:
                 The default uses CUDA stream object of the current context.
             order ({'C', 'F', 'A'}): The desired memory layout of the host
                 array. When ``order`` is 'A', it uses 'F' if the array is
-                fortran-contiguous and 'C' otherwise.
+                fortran-contiguous and 'C' otherwise. The ``order`` will be
+                ignored if ``out`` is specified.
+            out Output array. It should be a pinned memory to enable
```
Comment (niboshi) [inline cupy/core/core.pyx]: Please fix markup. We should specify the type as `numpy.ndarray`.

The sentence should be something like this:
"In order to enable asynchronous copy, the underlying memory should be a pinned memory."

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by jeng1220** [comment]
One suggestion that we need to modify some examples to educate user how to achieve asynchronous copy  (overlap data transfers) in correct way.
For example: https://github.com/cupy/cupy/blob/master/examples/stream/cupy_memcpy.py
Replace `x_pinned_cpu = x_gpu.get()` with `x_gpu.get(out=x_pinned_cpu)`

**Pull request merged**